Some objectGUID buffer cause error in escape filter function

ciePIC opened the pull request:
Search object by objectGUID in MS AD cause error for some GUID buffer.
For example with 2 real GUID :
- <Buffer e4 80 74 d3 3e 27 9c 46 b3 33 a3 5f 07 ff 87 a3> works
- <Buffer 44 29 12 42 1d e1 91 41 98 cc 40 92 51 52 7b df> cause error :
`TypeError: Cannot read properties of undefined (reading 'toString')`

This is caused because GUID buffer is not representing character and sometimes, escape function try to escape 2 or 3 bytes at the last buffer byte.
Fix by checking buffer index before try to add second/third byte.

ciePIC committed:
HOTFIX : objectGUID Buffer cause error

HOTFIX : objectGUID Buffer is limited to 16 bytes. When escape function need to escape a 2 or 3byte code, if index is at the end of the buffer, error are throw because trying to access unavailable index.
Fix by adding check of index position.

ciePIC committed:
Clean log

jsumners commented on lib/utils/escape-filter-value.js:
```diff
@@ -31,18 +31,17 @@ function escapeBuffer (buf) {
       // Represents the first byte in a 2-byte UTF-8 character.
-      result += '\\' + buf[i].toString(16) + '\\' + buf[i + 1].toString(16)
+      result += '\\' + buf[i].toString(16);
+      if(i<15) { result += '\\' + buf[i + 1].toString(16); }  //Only add second byte if exist in buffer
```
Where does the magic `15` constant come from?

ciePIC replied:
Thank for the question. It's a mistake. I was locked in my use case sorry.
I have commit change to replace this by "buf.length-1". Because if not buf[i+1] is undefined so .toString function cause issue.

jsumners commented on lib/utils/escape-filter-value.js:
```diff
@@ -31,18 +31,17 @@ function escapeBuffer (buf) {
-      ].join('')
+      result += '\\' + buf[i].toString(16);
+      if(i<15) { result += '\\' + buf[i + 1].toString(16); }  //Only add second byte if exist in buffer
+      if(i<14) { result += '\\' + buf[i + 2].toString(16); }  //Only add third byte if exist in buffer
```
Again, where does this `14` constant come from?

ciePIC replied:
Same as above. Sorry.
I have commit change to replace this by "buf.length-1" and "buf.length-2".

jsumners reviewed (changes requested):
Can you please provide an example of what causes the issue you are attempting to solve here? Also, we will need unit tests that cover these cases.

ciePIC committed:
Replace magic number

Replace magic number by number based on buffer length

ciePIC commented:
I try to retrive one user by his objectGUID from an Microsoft Active Directory with the simplified code as follow and ldap.js lib :

I pass the objectGUID buffer as value field of an EqualityFilter. For some user, it works very well but for some other, I get an error : `TypeError: Cannot read properties of undefined (reading 'toString')`
After investigating, this error is in the escape function. objectGUID is a 16 bytes buffer of binary data so, for example, the last byte can trig in the escape function of 2 or 3 bytes encoding. In this case the function was trying to access to the "i+1" byte but if it occur on the last buffer byte, there is no "+1 byte" so "undefined" is returned. And .toString function cannot be called on undefined value.
This case can append on every binary data.

I will think about implementing a unit test for this case because we can produce many and many use case....

````
//Two objectGUID retrive before from AD.
//First work as expected
//Second cause error
let myBuf1 = Buffer.from([0xe4, 0x80, 0x74, 0xd3, 0x3e, 0x27, 0x9c, 0x46, 0xb3, 0x33, 0xa3, 0x5f, 0x07, 0xff, 0x87, 0xa3]);
let myBuf2 = Buffer.from([0x44, 0x29, 0x12, 0x42, 0x1d, 0xe1, 0x91, 0x41, 0x98, 0xcc, 0x40, 0x92, 0x51, 0x52, 0x7b, 0xdf]);

//Create filter to retrive user
let filter = new filters.AndFilter({
    filters: [
        new filters.EqualityFilter({
            attribute: "objectclass",
            value: "user"
        }),
        new filters.EqualityFilter({
            attribute: "objectGUID",
            value: myBuf2
         })
    ]
});

//Search user
let ldapClient = createAndBindClient(); //Personnal function that use ldapjs and return a binded client with read authorizations.
let searchOptions = { filter: filter, scope: 'sub' };
ldapClient.search(searchBase, searchOptions, function (err, res) {
    var users = [];
    if (err) { reject(err); ldapClient.unbind(); return; }
    res.on('searchEntry', function (entry) { users.push(entry); });
    res.on('searchReference', function (referral) { /*NOT IMPLEMENTED */ });
    res.on('error', function (err) { reject(err); ldapClient.unbind(); });
    res.on('end', function (result) {
        if (result.status != 0) { reject(new Error('ldap search status is not 0, search failed')); }
        else { resolve(users); }
        ldapClient.unbind();
    })
})
````

ciePIC committed:
add unit-test

Add test for binary data ending by value that trig 2 or 3 bytes escaping UTF8. Implemented with a buffer length of 4bytes.

ciePIC commented:
Hello, I have add a unit test for my usecase.
I have write a simple test with low length Buffer that cause issue for the same reasons.

jsumners commented on lib/utils/escape-filter-value.js:
```diff
@@ -27,22 +27,22 @@ function escapeFilterValue (toEscape) {
 
 function escapeBuffer (buf) {
-  let result = ''
+  let bufLen = buf.length;
```
Please use `.byteLength` to be sure we account for multi-byte characters:
```suggestion
  let bufLen = buf.byteLength;
```

ciePIC replied:
OK, I will change it.

jsumners commented on lib/utils/escape-filter-value.js:
```diff
@@ -27,22 +27,22 @@ function escapeFilterValue (toEscape) {
       // Represents the first byte in a 2-byte UTF-8 character.
-      result += '\\' + buf[i].toString(16) + '\\' + buf[i + 1].toString(16)
+      result += '\\' + buf[i].toString(16);
+      if(i<bufLen-1) { result += '\\' + buf[i + 1].toString(16); }  //Only add second byte if exist in buffer
```
This is incomplete (the three byte condition is also incomplete):

```suggestion
      if (i < bufLen-1) {
        result += '\\' + buf[i + 1].toString(16)
        i += 2
        continue
      }
```

ciePIC replied:
Hello, Why it's incomplete ?
Also i increment and continu does not need to be in the if statment because the if is not executed only at end of the buffer so there is no more iteration in the for loop.

Also, I confirm that this code works perfectly.

jsumners commented on lib/utils/escape-filter-value.test.js:
```diff
@@ -83,3 +83,9 @@ tap.test('leaves encoded characters intact', async t => {
   }
 })
+
+tap.test('parse binary data with end byte >0xc0', async t => {
+  const input = Buffer.from([0x30, 0x41, 0x62, 0xd2])
+  const expected = '0Ab\\d2'
+  t.equal(escapeFilterValue(input), expected)
+})
```
```suggestion
tap.test('handle start of two byte utf-8 character at end of buffer', async t => {
  const input = Buffer.from([0x7e, 0xc2])
  const expected = '~\\dc2'
  t.equal(escapeFilterValue(input), expected)
})
```

We are also missing a test for the three byte case.

ciePIC replied:
OK, I will split into 2 more specific test

jsumners reviewed (commented):
Please run `npm run lint` prior to committing your changes.

ciePIC committed:
Fix eslint + byteLength + split test

ciePIC commented:
Hello, I have done changes. I wait answer about incomplete code because I don't understand...
